File: c#/Days/Day3.cs
using System.Text.RegularExpressions;

namespace aoc24.Days;

public static class Day3
{
    public static int Part1(string s)
    {
        var re = new Regex(@"mul\((?<n1>\d+),(?<n2>\d+)\)");
        var matches = re.Matches(s);
        var sol = 0;

        foreach (Match match in matches)
        {
            var n1 = match.Groups["n1"].Value;
            var n2 = match.Groups["n2"].Value;
            sol += int.Parse(n1) * int.Parse(n2);
        }

        return sol;
    }

    public static int Part2(string s)
    {
        var re = new Regex(@"(?<mul>mul\((?<n1>\d+),(?<n2>\d+)\))|(?<do>do\(\))|(?<dont>don't\(\))");
        var matches = re.Matches(s);
        var sol = 0;
        var on = true;

        foreach (Match match in matches)
        {
            if (match.Groups["mul"].Success && on)
            {
                var n1 = match.Groups["n1"].Value;
                var n2 = match.Groups["n2"].Value;
                sol += int.Parse(n1) * int.Parse(n2);
            }
            else if (match.Groups["do"].Success)
            {
                on = true;
            }
            else if (match.Groups["dont"].Success)
            {
                on = false;
            }
        }

        return sol;
    }
}

File: rust/src/days/day3.rs
use regex::Regex;

pub fn part1(input: &str) -> usize {
    let re = Regex::new(r"mul\((?<n1>\d+),(?<n2>\d+)\)").unwrap();
    let mut sol = 0;
    for m in re.captures_iter(input) {
        let n1 = m.name("n1").unwrap().as_str().parse::<usize>().unwrap();
        let n2 = m.name("n2").unwrap().as_str().parse::<usize>().unwrap();
        sol += n1 * n2;
    }
    sol
}

pub fn part2(input: &str) -> usize {
    let re = Regex::new(r"(?<mul>mul\((?<n1>\d+),(?<n2>\d+)\))|(?<do>do\(\))|(?<dont>don't\(\))").unwrap();
    let mut sol = 0;
    let mut on = true;
    for m in re.captures_iter(input) {
        if let Some(_) = m.name("mul") {
            if !on { continue }
            let n1 = m.name("n1").unwrap().as_str().parse::<usize>().unwrap();
            let n2 = m.name("n2").unwrap().as_str().parse::<usize>().unwrap();
            sol += n1 * n2;
        } else if let Some(_) = m.name("do") {
            on = true;
        } else if let Some(_) = m.name("dont") {
            on = false;
        }
    }
    sol
}

#[cfg(test)]
mod tests {
    use super::*;

    const INPUT: &str =
        r"xmul(2,4)&mul[3,7]!^don't()_mul(5,5)+mul(32,64](mul(11,8)undo()?mul(8,5))";

    #[test]
    fn test_part1() {
        assert_eq!(161, part1(INPUT));
    }

    #[test]
    fn test_part2() {
        assert_eq!(48, part2(INPUT));
    }
}
